mp2-common/src/eth.rs

```diff
@@ -111,6 +116,177 @@ pub struct ProofQuery {
+        let address_header = rlp_log_no_header.payload_info()?;
+        let rel_address_offset = log_header.header_len + address_header.header_len;
+        // Find the signature offset (skipping its header)
+        let topics_data = &buf[rel_address_offset + address_header.value_len
+            ..log_header.header_len + log_header.value_len];
+        let topics_rlp = rlp::Rlp::new(topics_data);
+        let topics_header = topics_rlp.payload_info()?;
+        let topic_0_data =
+            &buf[rel_address_offset + address_header.value_len + topics_header.header_len
+                ..log_header.header_len
+                    + address_header.header_len
+                    + address_header.value_len
+                    + topics_header.header_len
+                    + topics_header.value_len];
+        let topic_0_rlp = rlp::Rlp::new(topic_0_data);
+        let topic_0_header = topic_0_rlp.payload_info()?;
+        let rel_sig_offset = log_header.header_len
+            + address_header.header_len
+            + address_header.value_len
+            + topics_header.header_len
+            + topic_0_header.header_len;
+        let event_signature: [u8; 32] = buf[rel_sig_offset..rel_sig_offset + 32].try_into()?;
+        // Each topic takes 33 bytes to encode so we divide this length by 33 to get the number of topics remaining
+        let remaining_topics = buf[rel_sig_offset + topic_0_header.value_len
+            ..log_header.header_len
+                + address_header.header_len
+                + address_header.value_len
+                + topics_header.header_len
+                + topics_header.value_len]
+            .len()
+            / 33;
+
+        let mut topics = [LogDataInfo::default(); 3];
+        let mut current_topic_offset = rel_sig_offset + topic_0_header.value_len + 1;
+        topics
+            .iter_mut()
+            .enumerate()
+            .take(remaining_topics)
+            .for_each(|(j, info)| {
+                *info = LogDataInfo {
+                    column_id: j + 2,
+                    rel_byte_offset: current_topic_offset,
+                    len: 32,
+                };
+                current_topic_offset += 33;
+            });
+
+        // Deal with any remaining data
+        let mut data = [LogDataInfo::default(); 2];
+
+        let data_vec = if current_topic_offset < buf.len() {
+            buf.iter()
+                .skip(current_topic_offset - 1)
+                .copied()
+                .collect::<Vec<u8>>()
+        } else {
+            vec![]
+        };
+
+        if !data_vec.is_empty() {
+            let data_rlp = rlp::Rlp::new(&data_vec);
+            let data_header = data_rlp.payload_info()?;
+            // Since we can deal with at most two words of additional data we only need to take 66 bytes from this list
+            let mut additional_offset = data_header.header_len;
+            data_vec[data_header.header_len..]
+                .chunks(33)
```

Comment:
Yes it would, I wrote this part before getting used to how the `rlp` crate worked and then didn't come back to it but that seems like a much safer way of writing the code.